File: src/devices/device-provisioning/Business/Implements/ProvisioningBusiness.cs
using System;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace DeviceProvisioning.Business.Implements
{
    public class ProvisioningBusiness
    {
        private readonly ConfigurationFileAccess fileAccessor = new ConfigurationFileAccess();

        private bool IsDeviceIdFromConfigIsMatching()
        {
            var ownDeviceId = ReadDeviceId();
            var configuration = fileAccessor.GetConfiguration();
            return ownDeviceId != configuration.DeviceId;
        }

        public void SetupDevice()
        {

        }

        // If file exists, it means that device is already registered and nothing is needed
        public bool IsAlreadyRegistered
        {
            get
            {
                if (!fileAccessor.IsAlreadyRegistered)
                {
                    return false;
                }
                else
                {
                    return IsDeviceIdFromConfigIsMatching();
                };
            }
        }

        private static string ReadDeviceId()
        {
            var deviceInfos = File.ReadAllText("/proc/cpuinfo");
            var match = Regex.Match(deviceInfos, @"Serial\s*:\s([0-9a-f]{16})");

            if (match.Success)
            {
                return match.Value.Replace(" ", "")
                                             .Split(":").LastOrDefault();
            }
            
            throw new InvalidOperationException("Impossible to extract serial number from /proc/cpuinfo file");
        }
    }
}

File: src/devices/device-provisioning-node/Business/ProvisioningBusiness.ts
import { ConfigurationFileAccess } from "./ConfigurationFileAccess";
import { ProvisioningGenerator, ProvisioningGeneratorResultType } from "./ProvisionningGenerator";
import * as fs from "fs";
import { ConfigurationDevice } from "../Models/ConfigurationDevice";
import { ProvisioningDeviceClient } from 'azure-iot-provisioning-device';
import { ProvisioningFacade } from "./ProvisioningFacade";

export class ProvisioningBusiness {

    private fileAccessor = new ConfigurationFileAccess();
    private provisioningGenerator = new ProvisioningGenerator();
    private generatedDeviceId = this.ReadDeviceId();

    public SetupDevice(): void {
        const result = this.provisioningGenerator.Generate(this.generatedDeviceId);
        if (result.Type !== ProvisioningGeneratorResultType.Ok) {
            const error = `Cannot generate info for configuration. type : ${result.Type}`;
            throw error;
        }
        const configuration: ConfigurationDevice = {
            DeviceId: this.generatedDeviceId,
            GlobalDeviceEndpoint: result.GlobalDeviceEndpoint ?? "",
            IdScope: result.IdScope ?? "",
            PrimaryKey: result.PrimaryKey ?? "",
            SecondaryKey: result.SecondaryKey ?? "",
            IsRegistered: false,
        };
        this.fileAccessor.StoreConfiguration(configuration);

        const facade = new ProvisioningFacade();
        facade.register(configuration);
    }

    // // If file exists, it means that device is already registered and nothing is needed
    IsAlreadyRegistered(): boolean {
        if (!this.fileAccessor.IsAlreadyRegistered) {
            return false;
        } else {
            return this.IsDeviceIdFromConfigIsMatching();
        }
    }

    private IsDeviceIdFromConfigIsMatching(): boolean {
        const configuration = this.fileAccessor.GetConfiguration();
        return this.generatedDeviceId != configuration?.DeviceId;
    }

    private ReadDeviceId(): string {
        const deviceInfos = fs.readFileSync("/proc/cpuinfo", "utf8");
        if (!deviceInfos) {
            throw "Impossible to extract serial number from /proc/cpuinfo file";
        }
        const reg = new RegExp('Serial\\s*:\\s([0-9a-f]{16})');
        const results = deviceInfos.match(reg);

        if (results) {
            const firstResult = results[0];
            const deviceId = firstResult
                .replace(" ", "")
                .split(":")[1];
            console.log("deviceid: " + deviceId);
            return deviceId;
        } else {
            throw "No regex match";
        }

    }
}
